Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeLibWrapper.cs

```diff
@@ -0,0 +1,281 @@
+        private IntPtr _target_ITypeLibPtr;
+        private ITypeLibInternal _target_ITypeLib;
+        private bool _target_ITypeLib_IsRefCounted;
+
+        public bool HasVBEExtensions { get; private set; }
+        public TypeInfoWrapperCollection TypeInfos { get; private set; }
+
+        // helpers
+        public string Name => CachedTextFields._name;
+        public string DocString => CachedTextFields._docString;
+        public int HelpContext => CachedTextFields._helpContext;
+        public string HelpFile => CachedTextFields._helpFile;
+        public int TypesCount => _target_ITypeLib.GetTypeInfoCount();
+
+        public struct TypeLibTextFields
+        {
+            public string _name;
+            public string _docString;
+            public int _helpContext;
+            public string _helpFile;
+        }
+        private TypeLibTextFields? _cachedTextFields;
+        private TypeLibTextFields CachedTextFields
+        {
+            get
+            {
+                if (!_cachedTextFields.HasValue)
+                {
+                    var cache = new TypeLibTextFields();
+                    // as a C# caller, it's easier to work with ComTypes.ITypeLib
+                    ((ComTypes.ITypeLib)_target_ITypeLib).GetDocumentation((int)KnownDispatchMemberIDs.MEMBERID_NIL, out cache._name, out cache._docString, out cache._helpContext, out cache._helpFile);
+                    if (cache._name == null) cache._name = "[VBA.Immediate.Window]";
+                    _cachedTextFields = cache;
+                }
+                return _cachedTextFields.Value;
+            }
+        }
+
+        TypeLibVBEExtensions _vbeExtensions;
+        public TypeLibVBEExtensions VBEExtensions
+        {
+            get
+            {
+                if (_vbeExtensions == null)
+                {
+                    if (!HasVBEExtensions) throw new InvalidOperationException("This TypeLib does not represent a VBE project, so does not expose VBE Extensions");
+                    _vbeExtensions = new TypeLibVBEExtensions(this, _target_ITypeLib);
+                }
+
+                return _vbeExtensions;
+            }
+        }
+
+        private void InitCommon()
+        {
+            TypeInfos = new TypeInfoWrapperCollection(this);
+            HasVBEExtensions = (_target_ITypeLib as IVBEProject) != null;
+        }
+
+        internal static TypeLibWrapper FromVBProject(IVBProject vbProject)
+        {
+            using (var references = vbProject.References)
+            {
+                // Now we've got the references object, we can read the internal object structure to grab the ITypeLib
+                var internalReferencesObj = StructHelper.ReadComObjectStructure<VBEReferencesObj>(references.Target);
+                return new TypeLibWrapper(internalReferencesObj._typeLib, makeCopyOfReference: false);
+            }
+        }
+
+        private void InitFromRawPointer(IntPtr rawObjectPtr, bool makeCopyOfReference)
```

Comment:
I have my doubts about the procedure. This is a new procedure introduced and is used in the ctor of the class. See [this diff] to compare how it was originally implemented, in particular, we changed the parameter from `isRefCountedInput` to `makeCopyOfReference`:

https://github.com/rubberduck-vba/Rubberduck/blob/5c244248ef9f051d8048b3454d7cae7ddab6756d/Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeLibs.cs#L191-L200

The problem is that we do `GetObjectForIUnknown` which implicitly does an `AddRef`, so we are actually adding two references on the pointer. If we want to make a copy, we probably could remove the `AddRef` do `Release` right after `GetObjectForIUnknown`, but that would make the class potentially unsafe because it may then end up disconnected.

Reply:
Just a couple of quick comments.  This class does indeed take two independent references to the same underlying object... one intended for IntPtr/raw access (_target_ITypeLibPtr), and one for normal interop (_target_ITypeLib).    Most probably you could drop the _target_ITypeLibPtr stored version, since it looks like it is not actually being used for that purpose anymore, but I haven't looked in detail.

Be careful with regards to the makeCopyOfReference argument...  sometimes the constructor is being passed an object pointer to the ITypeLib interface that was obtained via unorthodox means (via a direct memory read of the VBE references object private field).  Ultimately, this means you've got a live object pointer but you haven't increased the reference count on said object yet.  Arguably, the problem of having a non-reference-counted object is not the responsibility of this class though, and so the manual correction to the reference count might be better refactored out to the caller for clarity.   

Good luck :)

Reply:
In the end, the implementation is fine - we did remove a reference to `TypeLibWrapper` from the `TypeLibVBEExtension` class which may or may not have helped.